enhancements/rhcos/split-rhcos-into-layers.md

```diff
@@ -0,0 +1,458 @@
+
+Another important follow-up would to be to hook up better CI testing to openshift/os. There is
+currently no CI test on that repo which actually launches a cluster. The reason is that RHCOS
+is just built differently. But now that the node image is simply yet another layered image build,
+it fits perfectly in Prow's opinionated model of building the image and shoving it in the test
+release payload with which to launch test cluster.
+
+#### Re-organize MCO templates
+
+Currently, there's a pile of templated files that live in the MCO. These
+files configure systemd services and run extra logic as required by OCP and
+conceptually have nothing to do with the MCO itself. It would then make sense to
+move them to the openshift/os repo alongside the rest of the definition for the
+node image.
+
+In a first step, we would leave them in the MCO but work on "de-templatizing"
+them by leveraging existing systemd directives as much as possible (e.g.
+conditionalizing on platform can be done with `ignition.platform.id=azure`,
+conditionalizing on OVN could be done with a generator or by having it pulled in
+by another systemd unit, etc...).
+
+Once they're fully de-templatized, we can move them to openshift/os so that they
+are part of the node image proper.
+
+This would considerable shrink the amount of templates that the MCO would
+need to process, and as a result serve via the MCS and/or layer via on-cluster
+layering.
+
+### Risks and Mitigations
+
+TODO
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly. For
+example, consider both security and how this will impact the larger OKD
+ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside your immediate sub-project.
+-->
+
+### Drawbacks
+
+- This will make it somewhat harder (or at the very least much more explicit) to
+  fast-track RHEL fixes into specific OCP releases. Because everything beneath
+  the node layer is RHEL-versioned and possibly shared across multiple OCP
+  releases, any fast-tracks would need to happen by doing an RPM upgrade within
+  the node layer. This is more cumbersome than just tagging packages into the
+  the ART plashets as currently done, but makes us more thorough in our known
+  deltas from RHEL.
+  - We should do this as a last resort and instead work more closely with RHEL
+    package maintainers to get fixes into the releases we need.
+
+## Open Questions [optional]
+
+- How should we adapt the bootstrapping process of the installer to handle the lack
+  of oc and kubelet in the bootimage? The easy approach would be to download and pivot
+  into the node image before continuing with bootstrapping, but that incurs an additional
+  reboot which is against the stated goals. Some other possibilities so far:
+  - Run the kubelet as a container
+  - Use podman to run the static pods
+  - Install the kubelet RPM
+  - `bootc apply-live`
```

Comment:
Would this result in running systemd services being stopped? If so this is a non-starter for ABI and Assisted.

Reply:
It wouldn't in this case, no. (Also I didn't mention it there, but this would be `rpm-ostree apply-live` for now; https://github.com/containers/bootc/issues/76).

But actually, I don't think this will work for the ISO case currently. Or at least, it'd need some work I think on the bootc/rpm-ostree side first.